File: java_desktop/src/test/java/FalseCrypt/Crypto/CryptoTest.java
package FalseCrypt.Crypto;
import java.io.IOException;
import java.nio.charset.Charset;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;

import org.junit.Assert;
import org.junit.jupiter.api.Test;

import FalseCrypt.Crypto.CryptoWrapper;
import FalseCrypt.Crypto.PasswordDeriveData;
import FalseCrypt.Crypto.WeakCryptoConfig;
import FalseCrypt.Crypto.WeakKeyGenerator;
import FalseCrypt.Crypto.WeakPasswordDerivation;

class CryptoTest {
	private static final byte[] salt = new byte[] {-26, 102, 102, -28, -72, 93, 90, -117};
	private static final byte[] key = new byte[] {8, 41, 99, -80, -51, -115, 68, 103};
	
	@Test
	public void SaltAlwaysTheSameStart() {
		byte[] testArray = WeakKeyGenerator.GenerateSalt();
		Assert.assertArrayEquals(testArray, salt);
	}
	
	@Test
	public void PasswordKeyOnceTheSame() throws NoSuchAlgorithmException, InvalidKeySpecException
    {
        PasswordDeriveData data = WeakPasswordDerivation.DerivePassword("Password");
        Assert.assertEquals(WeakCryptoConfig.SaltSizeBytes, data.Salt.length);
        Assert.assertEquals(WeakCryptoConfig.KeySizeBytes, data.Key.length);
        Assert.assertArrayEquals(salt, data.Salt);
        Assert.assertArrayEquals(key, data.Key);
    }

	@Test
	public void PasswordKeyAlwaysTheSame() throws NoSuchAlgorithmException, InvalidKeySpecException {
		for (int i = 0; i < 5; i++) {
			PasswordDeriveData data = WeakPasswordDerivation.DerivePassword("Password");
			Assert.assertEquals(WeakCryptoConfig.SaltSizeBytes, data.Salt.length);
			Assert.assertEquals(WeakCryptoConfig.KeySizeBytes, data.Key.length);
			Assert.assertArrayEquals(salt, data.Salt);
	        Assert.assertArrayEquals(key, data.Key);
		}
	}
	
//	@Test
	public void EncryptEncryptEqual() throws InvalidKeyException, NoSuchAlgorithmException, InvalidKeySpecException, NoSuchPaddingException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, IOException
    {
        final String message = "This is a secret Message";
        final String password = "password";

        String cipherText1 =
            CryptoWrapper.EncryptMessage(message, password, Charset.forName("UTF-8"));
        
        PasswordDeriveData data = WeakPasswordDerivation.DerivePassword(password);
        PasswordDeriveData data2 = WeakPasswordDerivation.DerivePassword(password);
        Assert.assertArrayEquals(data.Key, data2.Key);
        Assert.assertArrayEquals(data.Salt, data2.Salt);
        
        String cipherText2 = CryptoWrapper.EncryptMessage(message, data.Key, Charset.forName("UTF-8"));
        
//        System.out.println(CryptoWrapper.DecryptMessage(cipherText1, password, Charset.forName("UTF-8")));
//        System.out.println(CryptoWrapper.DecryptMessage(cipherText2, password, Charset.forName("UTF-8")));
        
        // strip IV / salt, not always present
        cipherText1 = cipherText1.substring(WeakCryptoConfig.SaltSizeBytes + WeakCryptoConfig.IV.length);
        cipherText2 = cipherText2.substring(WeakCryptoConfig.IV.length);
        Assert.assertEquals(cipherText1.length(), cipherText2.length());
        Assert.assertEquals(cipherText1, cipherText2);
    }
	
	@Test
	public void EncryptDecrypt() throws InvalidKeyException, NoSuchAlgorithmException, InvalidKeySpecException, NoSuchPaddingException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, IOException {
		String message = "This is a secret Message";
		String cipherText = CryptoWrapper.EncryptMessage(message, "password", Charset.forName("UTF-8"));
		String decryptedMessage = CryptoWrapper.DecryptMessage(cipherText, "password", Charset.forName("UTF-8"));
		Assert.assertEquals(message, decryptedMessage);
	}

	@Test
	public void EncryptDecryptManual() throws NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException, NoSuchPaddingException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, IOException {
		String message = "This is a secret Message";
		PasswordDeriveData data = WeakPasswordDerivation.DerivePassword("password");
		String cipherText = CryptoWrapper.EncryptMessage(message, data.Key, Charset.forName("UTF-8"));
		String decryptedMessage = CryptoWrapper.DecryptMessage(cipherText, data.Key, Charset.forName("UTF-8"));
		Assert.assertEquals(message, decryptedMessage);
	}
}

File: cs_desktop/tests/CryptoTest/CryptoTest.cs
using System.Linq;
using System.Text;
using FalseCrypt.Crypto;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace CryptoTest
{
    [TestClass]
    public class WeakCryptoTest
    {
        [TestMethod]
        public void SaltAlwaysTheSameStart()
        {
            var array = new byte[] {88, 144, 21, 224, 8, 102, 122, 218};
            var testArray = WeakKeyGenerator.GenerateSalt();
            Assert.AreEqual(true, testArray.SequenceEqual(array));
        }

        [TestMethod]
        public void PasswordKeyOnceTheSame()
        {
            var salt = new byte[] { 88, 144, 21, 224, 8, 102, 122, 218 };
            var key = new byte[] { 172, 19, 215, 234, 237, 198, 102, 232 };
            var data = WeakPasswordDerivation.DerivePassword("Password");

            Assert.AreEqual(WeakCryptoConfig.SaltSizeBytes, data.Salt.Length);
            Assert.AreEqual(WeakCryptoConfig.KeySizeBytes, data.Key.Length);
            Assert.AreEqual(true, data.Salt.SequenceEqual(salt));
            Assert.AreEqual(true, data.Key.SequenceEqual(key));
        }

        [TestMethod]
        public void PasswordKeyAlwaysTheSame()
        {
            var salt = new byte[] { 88, 144, 21, 224, 8, 102, 122, 218 };
            var key = new byte[] { 172, 19, 215, 234, 237, 198, 102, 232 };
            
            for (int i = 0; i < 5; i++)
            {
                var data = WeakPasswordDerivation.DerivePassword("Password");
                Assert.AreEqual(WeakCryptoConfig.SaltSizeBytes, data.Salt.Length);
                Assert.AreEqual(WeakCryptoConfig.KeySizeBytes, data.Key.Length);
                Assert.AreEqual(true, data.Salt.SequenceEqual(salt));
                Assert.AreEqual(true, data.Key.SequenceEqual(key));
            }
        }

        [TestMethod]
        public void EncryptEncryptEqual()
        {
            var message = "This is a secret Message";
            var password = "password";

            var cipherText1 =
                EncryptionCryptoWrapper.EncryptMessage(message, password, Encoding.UTF8);

            var data = WeakPasswordDerivation.DerivePassword(password);
            var cipherText2 = EncryptionCryptoWrapper.EncryptMessage(message, data.Key, Encoding.UTF8);

            Assert.AreNotEqual(cipherText1, cipherText2);
        }

        [TestMethod]
        public void EncryptDecrypt()
        {
            var message = "This is a secret Message";
            var cipherText =
                EncryptionCryptoWrapper.EncryptMessage(message, "password", Encoding.UTF8);
            var decryptedMessage = EncryptionCryptoWrapper.DecryptMessage(cipherText, "password", Encoding.UTF8);
            Assert.AreEqual(message, decryptedMessage);
        }

        [TestMethod]
        public void EncryptDecryptManual()
        {
            var message = "This is a secret Message";
            var data = WeakPasswordDerivation.DerivePassword("password");
            var cipherText = EncryptionCryptoWrapper.EncryptMessage(message, data.Key, Encoding.UTF8);
            var decryptedMessage = EncryptionCryptoWrapper.DecryptMessage(cipherText, data.Key, Encoding.UTF8);
            Assert.AreEqual(message, decryptedMessage);
        }
    }
}
